master/buildbot/worker/latent.py

```diff
@@ -236,7 +236,9 @@ def substantiate(self, wfb, build):
             return defer.succeed(False)
 
         if self.state == States.SUBSTANTIATED and self.conn is not None:
-            self._setBuildWaitTimer()
+            # Skip the wait timer if a build is already queued
+            if not self.building:
```

Comment:
This line is equivalent to deleting `self._setBuildWaitTimer()`. self.building checks whether at least one of WorkerForBuilder instances associated with the worker are busy (via `isBusy`). They will necessarily be busy, because at least one got its `buildStarted` called in the path to `Build.startBuild` which eventually indirectly calls `LatentWorker.substantiate()`.

However, this also means that `_setBuildWaitTimer` is completely unnecessary here. If `buildStarted` is guaranteed to be called when we are here, then regardless of what happens `buildFinished` will be called too for that particular build. That is the place for `_setBuildWaitTimer` which we already do there.

Reply:
All in all, thanks for the PR, it identified an important bug and did 95% of the work fixing it. I'll finish the polishing.

Reply:
Works for me, thanks. Do you think `self.building` / `isBusy` should be a part of the state machine? I found that lack of information in the worker's state to be terribly confusing to debug, which led to me exploring way too much of the call tree to find the root issue.

Reply:
`self.building` is quite separate from the internal state machine in that the states aren't dependent on each other. If we merge them then the state machine would grow double in size.

I agree that the latent worker itself is quite confusing. It became such once all edge cases got covered. It seems that there should be a way to refactor everything for simplicity, but nobody found any such way yet.